Libraries/LibWeb/Fetch/Fetching/FetchedDataReceiver.cpp

```diff
@@ -36,35 +37,89 @@ void FetchedDataReceiver::visit_edges(Visitor& visitor)
 
 void FetchedDataReceiver::set_pending_promise(GC::Ref<WebIDL::Promise> promise)
 {
-    auto had_pending_promise = m_pending_promise != nullptr;
+    VERIFY(!m_pending_promise);
+    VERIFY(!m_has_unfulfilled_promise);
     m_pending_promise = promise;
 
-    if (!had_pending_promise && !m_buffer.is_empty()) {
-        on_data_received(m_buffer);
-        m_buffer.clear();
+    if (!m_buffer.is_empty()) {
+        pull_bytes_into_stream(move(m_buffer));
+    } else if (m_lifecycle_state == LifecycleState::ReadyToClose) {
+        WebIDL::resolve_promise(m_stream->realm(), *m_pending_promise, JS::js_undefined());
+        m_pending_promise = {};
+        m_lifecycle_state = LifecycleState::Closed;
+        m_stream->close();
```

Comment:
Given that these 4 steps are repeated verbatim below, and that we're tracking quite a bit here - might be nicer to put this in a helper function